### PR TITLE
Added connection manager

### LINKED ISSUE #277: Connection Manager
A container for live PeerConnections. Other components can use this manager to add, remove and retrieve connections in a thread-safe way

### PR DESCRIPTION
**Pull request by sdbondi**
## Description

Connection manager allows other components to establish peer connections. 

## Motivation and Context

Closes #277 

## How Has This Been Tested?

Some basic tests have been written for components but much more tests need to be written

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [x] Breaking change (fix or feature that would cause existing functionality to change)
* [x] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [~] I have added tests to cover my changes.

### CODE REVIEW

File: comms/src/connection_manager/connections.rs [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+}
+
+impl LivePeerConnections {
+    /// Create a new live peer connection
+    pub fn new(context: Context, config: PeerConnectionConfig) -> Self {
+        Self {
+            context,
+            config,
+            repository: RwLock::new(ConnectionRepository::default()),
+            connection_thread_handles: RwLock::new(HashMap::new()),
```
Comment (neonknight64) [inline comms/src/connection_manager/connections.rs]: The `repository` and `connection_thread_handles` RwLocks might get out of sync, might be better to handle them using a single RwLock.

File: comms/src/connection_manager/protocol.rs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        &self,
+        connections: &Arc<LivePeerConnections>,
+        server_public_key: CurvePublicKey,
+    ) -> Result<EstablishedConnection>
+    {
+        let context = connections.borrow_context();
+        let config = &connections.config;
+
+        // TODO: Set net address stats and try all net addresses before giving up
+        let address = self.peer.addresses.addresses[0].clone().as_net_address();
```
Comment (neonknight64) [inline comms/src/connection_manager/protocol.rs]: I think you can keep calling `get_best_net_address` until it returns an error result, then all connection attempts would have been used

Comment (sdbondi) [reply comms/src/connection_manager/protocol.rs]: Problem is the mutable reference

File: comms/src/outbound_message_service/outbound_message_service.rs [hunk 3]
```diff
@@ -85,49 +80,44 @@ pub enum OutboundError {
 /// Handler functions use the OutboundMessageService to send messages to peers. The OutboundMessage service will receive
 /// messages from handlers, apply a broadcasting strategy, encrypted and serialized the messages into OutboundMessages
 /// and write them to the outbound message pool.
-pub struct OutboundMessageService<PubKey, SecKey, DS> {
```
Comment (neonknight64) [inline comms/src/outbound_message_service/outbound_message_service.rs]: Was removing the generics the right thing to do? It does make it easier having them hard coded in types.rs. Also, every module that needs the current nodes identity could be initialised with an Arc reference to the single copy of it.

File: comms/src/outbound_message_service/outbound_message_service.rs [hunk 4]
```diff
@@ -312,11 +292,11 @@ mod test {
             .chain(outbound_message.message_envelope.body_frame())
             .result()
             .to_vec();
-        assert!(signature.verify_challenge(&node_identity.public_key, &challenge));
+        assert!(signature.verify_challenge(&node_identity.identity.public_key, &challenge));
         // Check Encryption
         assert_eq!(message_envelope_header.flags, MessageFlags::ENCRYPTED);
         let ecdh_shared_secret =
-            RistrettoPublicKey::shared_secret(&dest_sk.clone(), &node_identity.public_key).to_vec();
+            RistrettoPublicKey::shared_secret(&node_identity.secret_key, &dest_peer.public_key).to_vec();
```
Comment (neonknight64) [inline comms/src/outbound_message_service/outbound_message_service.rs]: The new line is incorrect. This line tested that the destination node was able to generate the shared secret and decrypt the message.

File: comms/src/peer_manager/peer_manager.rs [hunk 5]
```diff
@@ -79,15 +79,15 @@ where
 
     /// Adds a peer to the routing table of the PeerManager if the peer does not already exist. When a peer already
     /// exist, the stored version will be replaced with the newly provided peer.
-    pub fn add_peer(&mut self, peer: Peer<PubKey>) -> Result<(), PeerManagerError> {
+    pub fn add_peer(&self, peer: Peer<PubKey>) -> Result<(), PeerManagerError> {
```
Comment (neonknight64) [inline comms/src/peer_manager/peer_manager.rs]: Good catch

File: comms/src/outbound_message_service/outbound_message_service.rs (hunk 4 above)
Comment (sdbondi) [reply comms/src/outbound_message_service/outbound_message_service.rs]: Ah ok, that will need to be worked on - the new PeerNodeIdentity the does not have the secret key (we wont have the destination peers secret key)

File: comms/src/connection_manager/protocol.rs (hunk 2 above)
Comment (neonknight64) [reply comms/src/connection_manager/protocol.rs]: PeerManager does expose the `get_best_net_address` and should be usable

File: comms/src/outbound_message_service/outbound_message_service.rs (hunk 4 above)
Comment (sdbondi) [reply comms/src/outbound_message_service/outbound_message_service.rs]: Ah ok, I've changed it to use the destination node's secret key - the new PeerNodeIdentity doesn't have a secret key (we won't know the dest node's secret key)

Comment (neonknight64) [reply comms/src/outbound_message_service/outbound_message_service.rs]: Perfect

File: comms/src/outbound_message_service/outbound_message_service.rs (hunk 3 above)
Comment (sdbondi) [reply comms/src/outbound_message_service/outbound_message_service.rs]: It made things a ton easier and is related to the static `NodeIdentity::global()` which could not use generics

**Pull request merged**